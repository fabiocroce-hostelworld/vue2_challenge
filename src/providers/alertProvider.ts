import { provide, inject, reactive, computed, ComputedRef } from "@vue/composition-api"

export enum alertTypes {
  SUCCESS = "success",
  ERROR = "error",
  INFO = "info",
  WARNING = "warning"
}

interface Alert {
  type: alertTypes
  message: string
  timerId?: number | undefined
}

export interface AlertProvider {
  addErrorAlert: (message: string) => void
  addSuccessAlert: (message: string) => void
  clearAlerts: () => void
  clearAlert: (idx: number) => void
  alerts: ComputedRef<Alert[]>
}

type alertConfig = { timeout?: number }

// Store for alert messages
export const setupAlertProvider = (config: alertConfig = {}): NonNullable<AlertProvider> => {
  const alertsArray = reactive<{ alerts: Array<Alert> }>({ alerts: [] })

  const clearByTimerId = (removeTimerId: number) => {
    alertsArray.alerts = alertsArray.alerts.filter(({ timerId }) => timerId !== removeTimerId)
  }

  const addAlert = ({ message, type }: Alert) => {
    if (config.timeout) {
      const timerId = setTimeout(() => {
        clearByTimerId(timerId)
      }, config.timeout)

      alertsArray.alerts.push({ message, type, timerId })
      return
    }
    alertsArray.alerts.push({ message, type })
  }

  const addErrorAlert = (message: string) => addAlert({ message, type: alertTypes.ERROR })
  const addSuccessAlert = (message: string) => addAlert({ message, type: alertTypes.SUCCESS })

  const clearAlerts = () => void (alertsArray.alerts.length = 0)
  const clearAlert = (removeIdx: number) => {
    alertsArray.alerts = alertsArray.alerts.filter((_, idx) => removeIdx !== idx)
  }
  const alerts = computed(() => alertsArray.alerts)

  return {
    addErrorAlert,
    addSuccessAlert,
    clearAlerts,
    clearAlert,
    alerts
  }
}

export const alertRef = Symbol("alert")

export const provideAlert = /* istanbul ignore next */ (config: alertConfig = {}) =>
  provide(alertRef, setupAlertProvider(config))
export const useAlert = () => inject<AlertProvider>(alertRef)
