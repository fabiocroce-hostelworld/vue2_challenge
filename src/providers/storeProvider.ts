import { provide, inject, computed, ComputedRef } from "@vue/composition-api"
import store from "@/store"

export interface StoreProvider {
  store: typeof store
  reactiveGetter: <T>(name: string) => ComputedRef<T>
}

// Vuex adapter for composition API
export const setupStoreProvider = (localStore: typeof store): StoreProvider => ({
  store: localStore,
  reactiveGetter: <T>(getterName: string): ComputedRef<T> => {
    return computed(() => localStore.getters[getterName])
  }
})

export const storeRef = Symbol("vuex-store")
export const provideStore = /* istanbul ignore next */ () => provide(storeRef, setupStoreProvider(store))
export const useStore = () => inject<StoreProvider>(storeRef)
