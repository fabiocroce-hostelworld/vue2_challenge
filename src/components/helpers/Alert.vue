<template>
  <div>
    <v-alert v-for="(alert, idx) in alerts" :key="idx" dense text :type="alert.type" @click="clearAlert(idx)" dismissible>
      {{ alert.message }}
    </v-alert>
  </div>
</template>

<script lang="ts">
import Vue from "vue"
import { AlertProvider, useAlert } from "@/providers/alertProvider"

export default Vue.extend({
  name: "Alert",
  setup() {
    const { alerts, clearAlert } = useAlert() as AlertProvider

    return {
      alerts,
      clearAlert
    }
  }
})
</script>
