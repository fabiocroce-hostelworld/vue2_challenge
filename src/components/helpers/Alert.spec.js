import { shallowMount, createLocalVue } from "@vue/test-utils"
import Alert from "@/components/helpers/Alert"
import { setupAlertProvider, alertRef, alertTypes } from "@/providers/alertProvider"

const alertProvider = setupAlertProvider({})
alertProvider.addSuccessAlert({ message: "Successful" })
alertProvider.addErrorAlert({ message: "Error" })

describe("Alert", () => {
  let input
  let localVue

  beforeEach(() => {
    localVue = createLocalVue()

    input = {
      localVue,
      provide: {
        [alertRef]: alertProvider
      }
    }
  })

  it("Snapshot test", () => {
    const wrapper = shallowMount(Alert, input)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("check for Alert component props", () => {
    const wrapper = shallowMount(Alert, input)

    const elements = wrapper.findAll(".alert")
    expect(elements.at(0).props().type).toBe(alertTypes.SUCCESS)

    expect(elements.at(1).props().type).toBe(alertTypes.ERROR)
  })

  it("Alert should disappear after being clicked", async () => {
    const wrapper = shallowMount(Alert, input)

    expect(alertProvider.alerts.value.length).toBe(2)

    await wrapper.find(".alert").vm.$emit("click")

    expect(alertProvider.alerts.value.length).toBe(1)
  })
})
