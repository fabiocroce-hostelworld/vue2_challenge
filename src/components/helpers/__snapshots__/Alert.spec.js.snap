// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Snapshot test 1`] = `
<div>
  <v-alert-stub tag="div"
                value="true"
                closelabel="$vuetify.close"
                dense="true"
                dismissible="true"
                closeicon="$cancel"
                icon
                text="true"
                type="success"
  >
    {
    "message": "Successful"
    }
  </v-alert-stub>
  <v-alert-stub tag="div"
                value="true"
                closelabel="$vuetify.close"
                dense="true"
                dismissible="true"
                closeicon="$cancel"
                icon
                text="true"
                type="error"
  >
    {
    "message": "Error"
    }
  </v-alert-stub>
</div>
`;
